### ISSUE TITLE
Two phase commit protocol is not respected when trying to commit a failed SQL transaction

### ISSUE THREAD

**Issue by sjustas** [psycopg/psycopg2#1213]
TL;DR: 
Situation: in two phase transaction, app executes an invalid statement SQL statment, then catches the exception, and finally, the application (wrongfully) tries to commit a failed transaction.  The perpared transaction is not created in tpc_preprare (but no error is raised!), and then tpc_commit or tpc_rollback fails with UndefinedObject, since transaction does not exist in postgres.

This breaks the two phase commit protocol, and leaves other databases in incosistent state.  The expectation is, that even if the app wrongfully tried to commit a failed SQL transaction, tpc_prepare would raise the error (voting phase), not tpc_commit/tpc_rollback (acknowledgment phase).

`UndefinedObject` error is also a bit weird, I would expect `InvalidTransactionState` or something similar.

------

I did a little debugging, and before proposing a PR, would like your feedback @dvarrazzo .

Postgres always returns PGRES_COMMAND_OK for PREPARE TRANSACTION, with one difference.  If transaction is in failed state, PQcmdStatus returns `ROLLBACK` instead of `PREPARE TRANSACTION` and transaction entry is not created in `pg_prepared_xacts` table:
```
In [6]: conn.tpc_prepare()
[1466019] pq_execute_command_locked: pgconn = 0x55b0ed3a7f60, query = PREPARE TRANSACTION '0_aWQtMQ==_Yi0x'
[1466019] pq_execute_command_locked pgstatus 1, PQcmdStatus: PREPARE TRANSACTION
```

```
In [13]: conn.tpc_prepare()
[1466019] pq_execute_command_locked: pgconn = 0x55b0ed3a7f60, query = PREPARE TRANSACTION '0_aWQtMg==_Yi0x'
[1466019] pq_execute_command_locked pgstatus 1, PQcmdStatus: ROLLBACK

```

So far the solution I could think of is basically doing a special case of `conn_tpc_command` for PREPARE TRANSACTION, that would `strcmp(PQcmdStatus(conn->pgres), "ROLLBACK")` and eventually raise `InvalidTransactionState`.

-----

-----
full debug output
-----

Tested with Postgres 12 and 9.6.

I've build psycopg2 with some additional debug ouput:
```
diff --git a/psycopg/pqpath.c b/psycopg/pqpath.c
index 52ba78b5..1c81d7db 100644
--- a/psycopg/pqpath.c
+++ b/psycopg/pqpath.c
@@ -285,9 +285,14 @@ pq_execute_command_locked(
         goto cleanup;
     }
 
+    Dprintf("pq_execute_command_locked pgstatus %d, PQcmdStatus: %s",
+            pgstatus, PQcmdStatus(conn->pgres));
     retvalue = 0;
     CLEARPGRES(conn->pgres);

```

Normal TPC operation.

```
In [3]: conn = psycopg2.connect(dsn)
In [4]: xid = conn.xid(0, "id-1", "b-1")
In [5]: conn.tpc_begin(xid)

In [6]: conn.tpc_prepare()
[1466019] conn_tpc_command: PREPARE TRANSACTION
[1466019] _pq_tpc_command: pgconn = 0x55b0ed3a7f60, command = PREPARE TRANSACTION
[1466019] pq_execute_command_locked: pgconn = 0x55b0ed3a7f60, query = PREPARE TRANSACTION '0_aWQtMQ==_Yi0x'
[1466019] pq_execute_command_locked pgstatus 1, PQcmdStatus: PREPARE TRANSACTION

In [7]: conn.tpc_rollback()
[1466019] conn_tpc_command: ROLLBACK PREPARED
[1466019] _pq_tpc_command: pgconn = 0x55b0ed3a7f60, command = ROLLBACK PREPARED
[1466019] pq_execute_command_locked: pgconn = 0x55b0ed3a7f60, query = ROLLBACK PREPARED '0_aWQtMQ==_Yi0x'
[1466019] pq_execute_command_locked pgstatus 1, PQcmdStatus: ROLLBACK PREPARED
```

TPC in a failed SQL transaction.

```
In [8]: xid = conn.xid(0, "id-2", "b-1")
In [9]: In [5]: conn.tpc_begin(xid)
In [10]: try:
    ...:     conn.cursor().execute("bad sql")
    ...: except Exception as err:
    ...:     print("Failure:", err)
    ...: 
Failure: syntax error at or near "bad"

In [12]: conn.status
2

In [13]: conn.tpc_prepare()
[1466019] conn_tpc_command: PREPARE TRANSACTION
[1466019] _pq_tpc_command: pgconn = 0x55b0ed3a7f60, command = PREPARE TRANSACTION
[1466019] pq_execute_command_locked: pgconn = 0x55b0ed3a7f60, query = PREPARE TRANSACTION '0_aWQtMg==_Yi0x'
[1466019] pq_execute_command_locked pgstatus 1, PQcmdStatus: ROLLBACK

In [14]: conn.status
5

In [15]: conn.tpc_rollback()
[1466019] conn_tpc_command: ROLLBACK PREPARED
[1466019] _pq_tpc_command: pgconn = 0x55b0ed3a7f60, command = ROLLBACK PREPARED
[1466019] pq_execute_command_locked: pgconn = 0x55b0ed3a7f60, query = ROLLBACK PREPARED '0_aWQtMg==_Yi0x'
[1466019] pq_execute_command_locked: result was not COMMAND_OK (7)
[1466019] pq_complete_error: pgconn = 0x55b0ed3a7f60, error = (null)
[1466019] pq_raise: PQresultErrorMessage: err=ERROR:  prepared transaction with identifier "0_aWQtMg==_Yi0x" does not exist

[1466019] pq_raise: err2=prepared transaction with identifier "0_aWQtMg==_Yi0x" does not exist

---------------------------------------------------------------------------
UndefinedObject                           Traceback (most recent call last)
<ipython-input-15-1cba975008b9> in <module>
----> 1 conn.tpc_rollback()

UndefinedObject: prepared transaction with identifier "0_aWQtMg==_Yi0x" does not exist
```

Raw SQL repro
```
postgres=# begin;
BEGIN
postgres=# prepare transaction 'one';
PREPARE TRANSACTION
postgres=# commit prepared 'one';
COMMIT PREPARED

postgres=# begin;
BEGIN
postgres=# bad sql;
ERROR:  syntax error at or near "bad"
LINE 1: bad sql;
        ^
postgres=# prepare transaction 'two';
ROLLBACK
postgres=# commit prepared 'two';
ERROR:  prepared transaction with identifier "two" does not exist
```

**Comment by sjustas** [comment]
Looking at drivers for other languages, it seems that no-one does any special handling ([Java/PgJDBC](https://github.com/pgjdbc/pgjdbc/blob/master/pgjdbc/src/main/java/org/postgresql/xa/PGXAConnection.java#L357), 
[.NET/Npgsql](https://github.com/npgsql/npgsql/blob/main/src/Npgsql/VolatileResourceManager.cs#L71))

So I guess it's best to be consistent.  Closing.
